starlette/routing.py

```diff
@@ -402,21 +410,23 @@ def routes(self) -> typing.List[BaseRoute]:
     def matches(self, scope: Scope) -> typing.Tuple[Match, Scope]:
         if scope["type"] in ("http", "websocket"):
             path = scope["path"]
-            match = self.path_regex.match(path)
+            root_path = scope.get("current_root_path", scope.get("root_path", ""))
+            current_path = scope.get("current_path", re.sub(r"^" + root_path, "", path))
+            match = self.path_regex.match(current_path)
             if match:
                 matched_params = match.groupdict()
                 for key, value in matched_params.items():
                     matched_params[key] = self.param_convertors[key].convert(value)
                 remaining_path = "/" + matched_params.pop("path")
-                matched_path = path[: -len(remaining_path)]
+                matched_path = current_path[: -len(remaining_path)]
                 path_params = dict(scope.get("path_params", {}))
                 path_params.update(matched_params)
                 root_path = scope.get("root_path", "")
                 child_scope = {
                     "path_params": path_params,
                     "app_root_path": scope.get("app_root_path", root_path),
-                    "root_path": root_path + matched_path,
-                    "path": remaining_path,
```

Comment:
Don't we need to keep the original `root_path` and `path` in the child scope?

Reply:
Aren't they always there anyway?

Reply:
Yes, you're right. My point was if we were using this `child_scope` as the new scope for inner routers, but it seems actually that it's always merged with the root `scope`.